feat(recording): Implement dropbox integration

hristoterezov opened the pull request:
Depends on https://github.com/jitsi/lib-jitsi-meet/pull/800

yanas commented on lang/main.json:
```diff
@@ -450,7 +450,12 @@
         "on": "Recording",
         "pending": "Preparing to record the meeting...",
         "rec": "REC",
+        "authDropboxText": "To start the recording you need to first authorize our Dropbox recording app. After the recording is finished the file will be uploaded to your Dropbox account.",
```
I would just replace this by "Upload your recording to Dropbox." and then add a "sign in" text to the button. I think the other part is kind of obvious.

yanas commented on lang/main.json:
```diff
@@ -450,7 +450,12 @@
         "pending": "Preparing to record the meeting...",
         "rec": "REC",
+        "authDropboxText": "To start the recording you need to first authorize our Dropbox recording app. After the recording is finished the file will be uploaded to your Dropbox account.",
+        "authDropboxCompletedText": "Our Dropbox app has been authorized successfully. You should see the recorded file in your Dropbox account shortly after the recording has finished.",
```
I would replace this by "Your recording file will appear in your Dropbox shortly after the recording has finished."

virtuacoplenny commented on react/features/base/oauth/reducer.js:
```diff
@@ -0,0 +1,38 @@
+ */
+PersistenceRegistry.register(STORE_NAME);
+
+ReducerRegistry.register('features/base/oauth',
```
Do you want to use STORE_NAME here?

hristoterezov replied:
Yes

virtuacoplenny commented on react/features/base/oauth/functions.js:
```diff
@@ -0,0 +1,29 @@
+ * @returns {Promise<string>} - The URL with the authorization details.
+ */
+export function authorize(authUrl: string): Promise<string> {
+    const windowName = `oauth${Date.now()}`;
```
Is there anything that prevents multiple auth windows for the same url from being opened? Does anything handle the user closing the dialog? For outlook, I had to add an interval to check if the window was closed while on the auth flow. I also stored the window so I could focus on it if the user tried to click the auth window again. This is one place where the divergence of implementations, which has I think been agreed on to be resolved after both features are merged, may be tricky.

hristoterezov replied:
Those use cases are not handled. I'm wondering if I need to handle them. 

On close it would be good to clean up the callback. But should we prevent the user for opening multiple windows?

PS: I'm not a big fan of polling. IMHO if this would be the cost to implement it, i'd rather leave those use cases unhandled.

virtuacoplenny commented on react/features/base/oauth/actions.js:
```diff
@@ -0,0 +1,49 @@
+                = parseURLParams(parseStandardURIString(returnUrl), true) || {};
+
+            dispatch(updateDropboxToken(params.access_token));
+        });
```
Can there be a .catch case here? Should there be?

hristoterezov replied:
The promise is never rejected. I don't think we should add catch.

virtuacoplenny commented on react/features/base/oauth/actions.js:
```diff
@@ -0,0 +1,49 @@
+
+        authorize(url).then(returnUrl => {
+            const params
+                = parseURLParams(parseStandardURIString(returnUrl), true) || {};
```
Does parseURLParams always return at least an empty object?

hristoterezov replied:
Yes

hristoterezov commented on react/features/recording/components/Recording/StartRecordingDialog.js:
```diff
@@ -0,0 +1,231 @@
+            appData: JSON.stringify({
+                'file_recording_metadata': {
+                    'upload_credentials': {
+                        'service_name': 'dropbox',
```
I guess when we add second service we will fix this problem. Don't you think it's better to leave it like it is for now instead of planning what would be the requirements for the second service and how to meet them?

hristoterezov commented on static/oauth.html:
```diff
@@ -0,0 +1,35 @@
+        }
+    </script>
+</head>
+<body />
```
Good to know!

hristoterezov commented on static/oauth.html:
```diff
@@ -0,0 +1,35 @@
+        if (typeof callback === 'function') {
+            callback();
+        } else {
+            alert('Something went wrong!');
```
Suggestions?

Maybe add the text in the HTML?

hristoterezov commented on static/oauth.html:
```diff
@@ -0,0 +1,35 @@
+    <!--#include virtual="/base.html" -->
+    <!--#include virtual="/title.html" -->
+    <script>
+        function getParentWindowCallback() {
```
OK.

hristoterezov commented on static/oauth.html:
```diff
@@ -0,0 +1,35 @@
+<html itemscope itemtype="http://schema.org/Product" prefix="og: http://ogp.me/ns#" xmlns="http://www.w3.org/1999/html">
```
Sure! The only problem is that we don't have a single html file with doctype. So let's be consistent here and fix them all in another PR?

hristoterezov commented on static/oauth.html:
```diff
@@ -0,0 +1,35 @@
+        }
+
+        var callback = getParentWindowCallback();
+        if (typeof callback === 'function') {
```
it's not double actually. this is checking the result of the function call only. how would you advice to change it?

hristoterezov replied:
It is but getParentWindowCallback also returns undefined. So what's your proposal for not doing the check? Do you want me to remove getParentWindowCallback?

hristoterezov commented on react/features/recording/components/Recording/StartRecordingDialogContent.web.js:
```diff
@@ -0,0 +1,189 @@
+                </div>
+                <div>{ t('recording.startRecordingBody') }</div>
+            </div>
+        );
```
Yes.

hristoterezov replied:
Yes. The toolbar recording button won't be displayed.

hristoterezov commented on react/features/recording/components/Recording/StartRecordingDialogContent.web.js:
```diff
@@ -0,0 +1,189 @@
+                        className = 'dropbox-logo'
+                        src = 'images/dropboxLogo.svg' />
+                </div>
+            </div>);
```
Do we have a rule for this?

hristoterezov replied:
I'm asking because I'm seeing some occurrences of parenthesis on the same line.

hristoterezov commented on react/features/recording/components/Recording/StartRecordingDialogContent.web.js:
```diff
@@ -0,0 +1,189 @@
+                <div>{ t('recording.authDropboxCompletedText') }</div>
+                <div className = 'logged-in-pannel'>
+                    <div>
+                        { t('recording.loggedIn', { userName }) }&nbsp;(&nbsp;
```
I haven't tried. Is this bad practice?

hristoterezov commented on react/features/recording/components/Recording/StartRecordingDialog.js:
```diff
@@ -0,0 +1,231 @@
+                isTokenValid: false,
+                isValidating: true
+            });
+            getDropboxData(_token, _clientId).then(data => {
```
My implementation doesn't use the error. IMHO it's the same.

hristoterezov commented on react/features/recording/components/Recording/StartRecordingDialog.js:
```diff
@@ -0,0 +1,231 @@
+                'file_recording_metadata': {
+                    'upload_credentials': {
+                        'service_name': 'dropbox',
+                        'token': _token
```
We want to record it. Why do you think it is security issue? It is stored in the local storage. If this is a security problem, then we have the same issue with a lot of other stuff that we put there.

hristoterezov commented on react/features/recording/components/Recording/StartRecordingDialogContent.web.js:
```diff
@@ -0,0 +1,189 @@
+                                spaceLeft,
+
+                                // assuming 1min -> 10MB recording:
+                                duration: Math.floor((spaceLeft || 0) / 10)
```
idk

hristoterezov commented:
@yanas @virtuacoplenny I think I addressed most of the comments. If I've missed something that you insist on let me know. Feel free to squash the commits before merging.
